sage/ip_lookup.py

```diff
@@ -0,0 +1,94 @@
+"""
+~40,000 distinct ip addresses recorded "downloading a file" in Q1 of 2024
+Using http://ip-api.com/batch API, I was able to extract the information of all ~40000 IP addresses.
+This information contains city, country, region, etc which is then stored back into snowflake
+for analysis
+"""
+from dotenv import dotenv_values
+import snowflake.connector
+from snowflake.connector.pandas_tools import write_pandas
+import pandas as pd
+import numpy as np
+
+import requests
+
+config = dotenv_values("../.env")
+
+ctx = snowflake.connector.connect(
+    user=config['user'],
+    password=config['password'],
+    account=config['snowflake_account'],
+    database="synapse_data_warehouse",
+    schema="synapse",
+    role="SYSADMIN",
+    warehouse="compute_xsmall"
+)
+cs = ctx.cursor()
+
+query = """
+select
+    distinct processedaccess.x_forwarded_for as unique_ips
+from
+    synapse_data_warehouse.synapse.filedownload
+inner join
+    synapse_data_warehouse.synapse.processedaccess
+    on filedownload.session_id = processedaccess.session_id
+where
+    filedownload.record_date between DATE('2024-01-01') and DATE('2024-03-31') and
+    processedaccess.record_date between DATE('2024-01-01') and DATE('2024-03-31');
+"""
+
+cs.execute(query)
+unique_ips = cs.fetch_pandas_all()
+import time
+
+
+def get_ip_info(ip_list: list):
+    while True:
+        try:
+            ip_info_response = requests.post('http://ip-api.com/batch', json=ip_list)
```

Comment:
It might be overengineering a solution to the problem, but if you wanted to start trying some async stuff here you could split these up and use
https://superfastpython.com/asyncio-gather/
https://www.python-httpx.org/async/

Reply:
Thanks @BryanFauble  - I'll explore when I have a chance.  What I'm dealing with here is an API rate limit of 15 per minute (the documentation says 45 per minute, but that's not what it is)

So doing an async would almost be too efficient, but educational